cryptography/src/handshake.rs

```diff
@@ -0,0 +1,246 @@
+use rand_core::CryptoRngCore;
+
+use crate::{
+    transcript::{Summary, Transcript},
+    PublicKey, Signer, Verifier,
+};
+
+const NAMESPACE: &'static [u8] = b"commonware/handshake";
+
+pub struct Msg1<S> {
+    time_ms: u64,
+    epk: x25519_dalek::PublicKey,
+    sig: S,
+}
+
+pub struct Msg2<S> {
+    time_ms: u64,
+    epk: x25519_dalek::PublicKey,
+    sig: S,
+    confirmation: Summary,
+}
+
+pub struct Msg3 {
+    confirmation: Summary,
+}
+
+pub struct DialState<P> {
+    esk: x25519_dalek::EphemeralSecret,
+    peer_identity: P,
+    transcript: Transcript,
+}
+
+pub struct ListenState {
+    confirmation: Summary,
+    send: SendCipher,
+    recv: RecvCipher,
+}
+
+struct CounterNonce {
+    inner: u128,
+}
+
+impl CounterNonce {
+    pub fn new() -> Self {
+        Self { inner: 0 }
+    }
+
+    pub fn inc(&mut self) -> [u8; 16] {
+        if self.inner >= 1 << 96 {
+            panic!("overflowed nonce");
+        }
+        let out = self.inner.to_le_bytes();
+        self.inner += 1;
+        out
+    }
+}
+
+pub struct SendCipher {
+    nonce: CounterNonce,
+    inner: ChaCha20Poly1305,
+}
+
+impl SendCipher {
+    fn new(mut rng: impl CryptoRngCore) -> Self {
+        let mut key = [0u8; 32];
+        rng.fill_bytes(&mut key[..]);
+        Self {
+            nonce: CounterNonce::new(),
+            inner: ChaCha20Poly1305::new(&key.into()),
+        }
+    }
+
+    pub fn send(&mut self, data: &[u8]) -> Vec<u8> {
+        self.inner
+            .encrypt((&self.nonce.inc()[..12]).into(), data)
+            .unwrap()
+    }
+}
+
+pub struct RecvCipher {
+    nonce: CounterNonce,
+    inner: ChaCha20Poly1305,
+}
+
+impl RecvCipher {
+    fn new(mut rng: impl CryptoRngCore) -> Self {
+        let mut key = [0u8; 32];
+        rng.fill_bytes(&mut key[..]);
+        Self {
+            nonce: CounterNonce::new(),
+            inner: ChaCha20Poly1305::new(&key.into()),
+        }
+    }
+
+    pub fn recv(&mut self, encrypted_data: &[u8]) -> Vec<u8> {
+        self.inner
+            .decrypt((&self.nonce.inc()[..12]).into(), encrypted_data)
+            .unwrap()
+    }
+}
+
+pub struct Context<S, P> {
+    current_time: u64,
+    my_identity: S,
+    peer_identity: P,
+}
+
+impl<S, P> Context<S, P> {
+    pub fn new(current_time_ms: u64, my_identity: S, peer_identity: P) -> Self {
+        Self {
+            current_time: current_time_ms,
+            my_identity,
+            peer_identity,
+        }
+    }
+}
+
+pub fn dial_start<S: Signer, P: PublicKey>(
+    rng: impl CryptoRngCore,
+    ctx: Context<S, P>,
+) -> (DialState<P>, Msg1<<S as Signer>::Signature>) {
+    let Context {
+        current_time,
+        my_identity,
+        peer_identity,
+    } = ctx;
+    let esk = x25519_dalek::EphemeralSecret::random_from_rng(rng);
+    let epk = x25519_dalek::PublicKey::from(&esk);
+    let mut transcript = Transcript::new(NAMESPACE);
+    let sig = transcript
+        .commit(peer_identity.encode())
```

Comment:
If we modify `commit` to take `Codec`, I think we could remove all these `encode`s?

Reply:
I suspect that is probably too aggressive but it would look noice

Reply:
Maybe we could offer a `CodecTranscript` that does this? Or add a `wrap-like` function?

https://github.com/commonwarexyz/monorepo/blob/bb428d2c8b9f94095d92c8cba59bbb196fc20412/p2p/src/utils/codec.rs#L7

Reply:
We could also add a wrapper method to transcript which takes something implementing `Encode`